### PR TITLE
Create Area Light Tool update and various pyside small changes, this update affects HiveUI so good to merge zooCore

### PR DESCRIPTION
**Pull request by andrewsilke**
- Create Area Light Tool in Toolsets has a minimised section and improved look
- renamed ComboBox, to ComboBoxSearchable which is more accurate, affects HiveUI and more
- fixed bug for area light not opening because of colored buttons
- layouts added CheckBoxRegular and updated comboBoxRegular functions for extra options in one line.  Updated "Create Area Light" to one liners for combobox and checkbox widgets
- added checkbox tick icons for better checkboxes
- added hardcoded bg colors for lineEdits and checboxes, needs to be stylseeted later
- add extra functionality to styled buttons, max and min width and height

### CODE REVIEW

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 1]
```diff
@@ -523,5 +528,166 @@ def keyReleaseEvent(self, event):
             self.ttKeyPressed = False
 
 
-
+class BtnTransparentBG(ExtendedButton):
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/extendedbutton.py]: Such a small override, why not just a function?

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 2]
```diff
@@ -523,5 +528,166 @@ def keyReleaseEvent(self, event):
...
+            self.setMinimumWidth(utils.dpiScale(minWidth))
+        if maxWidth:
+            self.setMaximumWidth(utils.dpiScale(maxWidth))
+        if minHeight:
+            self.setMinimumHeight(utils.dpiScale(minHeight))
+        if maxHeight:
+            self.setMaximumHeight(utils.dpiScale(maxHeight))
+
+
+class BtnRegular(QtWidgets.QPushButton):
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/extendedbutton.py]: same here

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 3]
```diff
@@ -523,5 +528,166 @@ def keyReleaseEvent(self, event):
...
+            self.setMinimumWidth(utils.dpiScale(minWidth))
+        if maxWidth:
+            self.setMaximumWidth(utils.dpiScale(maxWidth))
+        if minHeight:
+            self.setMinimumHeight(utils.dpiScale(minHeight))
+        if maxHeight:
+            self.setMaximumHeight(utils.dpiScale(maxHeight))
+
+
+def BtnStyle(text=None, icon=None, parent=None, toolTip="", style=BTN_DEFAULT, textCaps=False,
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/extendedbutton.py]: repeating arguments of the above classes? why not just use **kwargs and reference the docstrings of the other functions.
eg. 
`"""Create a button with text or an icon in various styles and options
     Style - 0 - BTN_DEFAULT - Default pyside button with optional text or an icon. See :func: BtnRegular for arguments.
    Style - 1 - BTN_TRANSPARENT_BG - Default pyside button w transparent bg. Icon colour is grey and lighter with hover See :func: BtnTransparentBG  for arguments.
"""`

File: zoo/libs/pyqt/widgets/layouts.py [hunk 4]
```diff
@@ -32,24 +31,30 @@ def __init__(self, label, placeholderText=None, buttonText=None, parent=None, ed
         :type btnRatio: float
         :param editRatio: the width ratio of the label/text/button corresponds to the ratios of labelRatio/btnRatio
         :type editRatio: float
+        :param toolTip: the tool tip message on mouse over hover, extra info
+        :type toolTip: str
         """
         super(StringEdit, self).__init__(parent=parent)
         self.edit = QtWidgets.QLineEdit(parent=self)
+        # hardcoding here as a temp workaround -silky
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: best to do a todo directive which by doing so will be easily searchable later .
`:todo: `

File: zoo/libs/pyqt/widgets/layouts.py [hunk 5]
```diff
@@ -84,15 +89,26 @@ def __init__(self, label="", items=(), parent=None):
...
         :type parent: class
         """
-        super(ComboBox, self).__init__(parent=parent)
+        super(ComboBoxSearchable, self).__init__(parent=parent)
         layout = QtWidgets.QHBoxLayout(self)
         layout.setContentsMargins(2, 2, 2, 2)
         layout.setSpacing(uiconstants.SPACING)
         self.box = combobox.ExtendedComboBox(items, parent)
 
         if label != "":
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: why not just `if label` ?

File: zoo/libs/pyqt/widgets/layouts.py [hunk 6]
```diff
@@ -71,10 +76,10 @@ def text(self):
         return self.edit.text()
 
 
-class ComboBox(QtWidgets.QWidget):
+class ComboBoxSearchable(QtWidgets.QWidget):
     itemChanged = QtCore.Signal(int, str)
 
-    def __init__(self, label="", items=(), parent=None):
+    def __init__(self, label="", items=(), parent=None, labelRatio=None, boxRatio=None, toolTip="", setIndex=0):
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: Am i correct in assuming that the ratio means the stretch value?

File: zoo/libs/pyqt/widgets/layouts.py [hunk 7]
```diff
@@ -153,6 +169,10 @@ def __init__(self, label="", items=None, parent=None, labelRatio=None, boxRatio=
         :type items: list
         :param parent: the qt parent
         :type parent: class
+        :param toolTip: the tooltip info to display with mouse hover
+        :type toolTip: str
+        :param setIndex: set the combo box value as an int - 0 is the first value, 1 is the second
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: Somewhat confusing description when comparing the logic of the code on line: 200. combo box value would suggest it's the value ie the label. 
maybe something more like:
:param setIndex: sets the combo box current index, default is 0

Comment (dsparrow27) [reply zoo/libs/pyqt/widgets/layouts.py]: also setIndex is weird naming. maybe just `index=0` or `currentIndex` or `defaultIndex` ?

File: zoo/libs/pyqt/widgets/layouts.py [hunk 8]
```diff
@@ -174,6 +194,10 @@ def __init__(self, label="", items=None, parent=None, labelRatio=None, boxRatio=
             layout.addWidget(self.box, boxRatio)
         else:
             layout.addWidget(self.box)
+        if toolTip:
+            self.box.setToolTip(toolTip)
+        if setIndex:
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: no need to check this. it's defaulted to zero, and Qt defaults to zero, just set it regardless.

File: zoo/libs/pyqt/widgets/layouts.py [hunk 9]
```diff
@@ -174,6 +194,10 @@ def __init__(self, label="", items=None, parent=None, labelRatio=None, boxRatio=
             layout.addWidget(self.box, boxRatio)
         else:
             layout.addWidget(self.box)
+        if toolTip:
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: Remove and just set the tooltip regardless.

File: zoo/libs/pyqt/widgets/layouts.py [hunk 10]
```diff
@@ -218,6 +242,34 @@ def setToText(self, text):
...
+    :param parent: the qt parent
+    :type parent: class
+    :param toolTip: the tooltip info to display with mouse hover
+    :type toolTip: str
+    :return qWidget: the checkbox qt widget
+    :rtype value: object
+    """
+    box = QtWidgets.QCheckBox(label, parent=parent)
+    # box = QtWidgets.QComboBox(label, parent=parent)
+    if toolTip:
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: same here

File: zoo/libs/pyqt/widgets/layouts.py [hunk 11]
```diff
@@ -218,6 +242,34 @@ def setToText(self, text):
...
+    :type setChecked: bool
+    :param parent: the qt parent
+    :type parent: class
+    :param toolTip: the tooltip info to display with mouse hover
+    :type toolTip: str
+    :return qWidget: the checkbox qt widget
+    :rtype value: object
+    """
+    box = QtWidgets.QCheckBox(label, parent=parent)
+    # box = QtWidgets.QComboBox(label, parent=parent)
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: remove comment out code

File: zoo/libs/pyqt/widgets/layouts.py [hunk 12]
```diff
@@ -218,6 +242,34 @@ def setToText(self, text):
...
+    if toolTip:
+        box.setToolTip(toolTip)
+    if setChecked:
+        box.setChecked(setChecked)
+    # hardcoding here as a temp workaround -silky
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/layouts.py]: :todo directive ?

File: zoo/libs/utils/colour.py [hunk 13]
```diff
@@ -307,4 +298,74 @@ def apply(self, r, g, b):
         rx = r * self.matrix[0][0] + g * self.matrix[0][1] + b * self.matrix[0][2]
         gx = r * self.matrix[1][0] + g * self.matrix[1][1] + b * self.matrix[1][2]
         bx = r * self.matrix[2][0] + g * self.matrix[2][1] + b * self.matrix[2][2]
-        return clamp(rx), clamp(gx), clamp(bx)
+        return zoomath.clamp(rx, 0, 255), zoomath.clamp(gx, 0, 255), zoomath.clamp(bx, 0, 255)
+
+
+def hslColourOffsetFloat(rgb, hueOffset=0, saturationOffset=0, lightnessOffset=0):
+    """Offset color with hue, saturation and lightness (brighten/darken) values
```
Comment (dsparrow27) [inline zoo/libs/utils/colour.py]: Put a space between this line and the below. 
Above line is the header and the below is the body.

File: zoo/libs/pyqt/widgets/extendedbutton.py (hunk 1 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: ah I did that from last time

File: zoo/libs/pyqt/widgets/extendedbutton.py (hunk 2 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: yep

File: zoo/libs/pyqt/widgets/extendedbutton.py (hunk 3 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: ah yeah good idea, I've never done that

File: zoo/libs/pyqt/widgets/layouts.py (hunk 4 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: that's why I put -silky so I could search :)  But that sounds better

File: zoo/libs/pyqt/widgets/layouts.py (hunk 5 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: that must be from a  long time ago.

File: zoo/libs/pyqt/widgets/layouts.py (hunk 9 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: sure

File: zoo/libs/pyqt/widgets/layouts.py (hunk 10 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: yrp

File: zoo/libs/pyqt/widgets/layouts.py (hunk 11 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: yep

File: zoo/libs/pyqt/widgets/layouts.py (hunk 12 above)
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/layouts.py]: yep

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 14]
```diff
@@ -523,5 +528,112 @@ def keyReleaseEvent(self, event):
...
+    btn = ExtendedButton(parent=kwargs['parent'], text=kwargs['text'])
+    if kwargs['icon']:
+        btn.setIconByName(kwargs['icon'], colors=kwargs['iconColor'])
+        """ todo: icon colorized anti aliasing is not working correctly?  Icons appear thicker
+        # will want this code later
+        self.setIcon(iconlib.iconColorized(icon, size=iconSize, color=iconColor, overlayName=overlayIconName,
+                           overlayColor=overlayIconColor))
+        """
+    btn.setToolTip(kwargs['toolTip'])
+    if kwargs['minWidth']:
```
[the remark this answers is not in the document]
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: done

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 15]
```diff
@@ -523,5 +528,112 @@ def keyReleaseEvent(self, event):
...
+    """
+    btn = ExtendedButton(parent=kwargs['parent'], text=kwargs['text'])
+    if kwargs['icon']:
+        btn.setIconByName(kwargs['icon'], colors=kwargs['iconColor'])
+        """ todo: icon colorized anti aliasing is not working correctly?  Icons appear thicker
+        # will want this code later
+        self.setIcon(iconlib.iconColorized(icon, size=iconSize, color=iconColor, overlayName=overlayIconName,
+                           overlayColor=overlayIconColor))
+        """
+    btn.setToolTip(kwargs['toolTip'])
```
[the remark this answers is not in the document]
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: done

File: zoo/libs/pyqt/widgets/extendedbutton.py [hunk 16]
```diff
@@ -523,5 +528,134 @@ def keyReleaseEvent(self, event):
...
+    :type kwargs: dict
+    :return qtBtn: returns a qt button widget
+    :rtype qtBtn: object
+    """
+    parent = kwargs.get("parent")
+    text = kwargs.get("text")
+    icon = kwargs.get("icon", (255, 255, 255))
+    minWidth = kwargs.get("minWidth")
+    toolTip = kwargs.get("toolTip", "")
+    iconColor = kwargs.get("iconColor")
```
[the remark this answers is not in the document]
Comment (andrewsilke) [reply zoo/libs/pyqt/widgets/extendedbutton.py]: left this one for the todo:

**Pull request merged**